File: src/test/java/dev/jlo/gamemodes/paper/CommandRouterTest.java
package dev.jlo.gamemodes.paper;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;
import java.util.UUID;
import org.junit.jupiter.api.Test;

class CommandRouterTest {
    @Test
    void joinStatusLeaveAndTeamRouteThroughService() {
        FakeCommandService service = new FakeCommandService();
        CommandRouter router = new CommandRouter(service);
        UUID player = UUID.randomUUID();

        assertEquals("joined OPR", router.execute(player, List.of("join", "opr")));
        assertEquals("team A", router.execute(player, List.of("team")));
        assertEquals("OPR", router.execute(player, List.of("status")));
        assertEquals("left", router.execute(player, List.of("leave")));
    }

    @Test
    void aliasesMapToJoinMode() {
        FakeCommandService service = new FakeCommandService();
        CommandRouter router = new CommandRouter(service);
        UUID player = UUID.randomUUID();

        assertEquals("joined OPR", router.execute(player, List.of("opr")));
        assertEquals("joined SIEGE", router.execute(player, List.of("siege")));
    }

    private static final class FakeCommandService implements CommandService {
        private Mode mode;
        private TeamAssignment team;

        @Override
        public TeamAssignment join(UUID player, Mode mode, String arena) {
            this.mode = mode;
            this.team = mode == Mode.OPR ? TeamAssignment.A : TeamAssignment.B;
            return team;
        }

        @Override
        public boolean leave(UUID player) {
            mode = null;
            team = null;
            return true;
        }

        @Override
        public boolean ready(UUID player) {
            return true;
        }

        @Override
        public String status(UUID player) {
            return mode == null ? "NONE" : mode.name();
        }

        @Override
        public TeamAssignment team(UUID player) {
            return team;
        }

        @Override
        public String admin(UUID player, String action, List<String> args) {
            return action;
        }
    }
}

File: src/test/kotlin/dev/jlo/gamemodes/paper/CommandRouterTest.kt
package dev.jlo.gamemodes.paper

import java.util.UUID
import kotlin.test.Test
import kotlin.test.assertEquals

class CommandRouterTest {
    @Test
    fun `join status leave and team route through service`() {
        val service = FakeCommandService()
        val router = CommandRouter(service)
        val player = UUID.randomUUID()

        assertEquals("joined OPR", router.execute(player, listOf("join", "opr")))
        assertEquals("team A", router.execute(player, listOf("team")))
        assertEquals("OPR", router.execute(player, listOf("status")))
        assertEquals("left", router.execute(player, listOf("leave")))
    }

    @Test
    fun `aliases map to join mode`() {
        val service = FakeCommandService()
        val router = CommandRouter(service)
        val player = UUID.randomUUID()

        assertEquals("joined OPR", router.execute(player, listOf("opr")))
        assertEquals("joined SIEGE", router.execute(player, listOf("siege")))
    }
}

private class FakeCommandService : CommandService {
    private var mode: Mode? = null
    private var team: TeamAssignment? = null
    override fun join(player: UUID, mode: Mode, arena: String?): TeamAssignment {
        this.mode = mode
        this.team = if (mode == Mode.OPR) TeamAssignment.A else TeamAssignment.B
        return team!!
    }
    override fun leave(player: UUID): Boolean { mode = null; team = null; return true }
    override fun ready(player: UUID): Boolean = true
    override fun status(player: UUID): String = mode?.name ?: "NONE"
    override fun team(player: UUID): TeamAssignment? = team
    override fun admin(player: UUID?, action: String, args: List<String>): String = action
}
